### PR TITLE
Deprecate interpolate

### PR DESCRIPTION
**Pull request by pp-mo**
Deprecation of ```cube.regridded``` and the whole of ```iris.analysis.interpolate``` module.
We basically want all of these older methods updated to the new scheme-based "cube.regrid" and "cube.interpolate".

It's a bit noisy, partly unavoidable due to renames.
It should be clearer from the individual commits.

### CODE REVIEW

File: lib/iris/_deprecation_helpers.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        # Create a wrapper init function which issues the deprecation.
+        def initfn(self, *args, **kwargs):
+            print(repr(depr_warnstring))
+            warnings.warn(depr_warnstring)
+            self._target_parent_class.__init__(self, *args, **kwargs)
+
+        # Set this as the init for the wrapper class.
+        initfn.func_name = '__init__'
+        # Also copy the original docstring.
+        initfn.__doc__ = parent_class.__init__.__doc__
```
Comment (rhattersley) [inline lib/iris/_deprecation_helpers.py]: Probably cleaner to do this with `functools.wraps`

Comment (pp-mo) [inline lib/iris/_deprecation_helpers.py]: I did wonder about that, but I couldn't quite work out how to make it go.
I'll take another look.

Comment (rhattersley) [inline lib/iris/_deprecation_helpers.py]: Perhaps...
```python
@wraps(parent_class.__init__)
def initfn(self, *args, **kwargs):
    print(...)
```

File: lib/iris/_deprecation_helpers.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        # Get a warning message from the underlying class.
+        depr_warnstring = class_dict.get('_DEPRECATION_WARNING')
+
+        # Save the parent class *on* the wrapper class, so we can chain to its
+        #  __init__ call.
+        class_dict['_target_parent_class'] = parent_class
+
+        # Create a wrapper init function which issues the deprecation.
+        def initfn(self, *args, **kwargs):
+            print(repr(depr_warnstring))
```
Comment (rhattersley) [inline lib/iris/_deprecation_helpers.py]: Is this supposed to be here?

File: lib/iris/tests/test_coding_standards.py [hunk 3]
```diff
@@ -73,6 +73,7 @@ class StandardReportWithExclusions(pep8.StandardReport):
     expected_bad_files = [
         '*/iris/std_names.py',
         '*/iris/analysis/interpolate.py',
```
Comment (rhattersley) [inline lib/iris/tests/test_coding_standards.py]: If you tidy up two small errors you can get rid of this line.
```
$ pep8 lib/iris/analysis/interpolate.py
lib/iris/analysis/interpolate.py:60:1: E302 expected 2 blank lines, found 1
lib/iris/analysis/interpolate.py:138:1: W391 blank line at end of file
```

**Comment by rhattersley** [comment]
Looks like excellent progress - thanks @pp-mo :smile: 

Good idea to keep using the old code internally so we don't have to transition our internal usage (with the likely impact on test results) in the same PR. :+1:

File: lib/iris/_deprecation_helpers.py (hunk 1 above)
Comment (pp-mo) [inline lib/iris/_deprecation_helpers.py]: ... actually, I've got a problem here...
The documentation is not getting the constructor signatures right for any of my classes wrapped with this 'ClassDeprecationWrapper' : They all appear to be (*args, **kwargs).
This is not helpful + not what I intended.
I see that the existing master already has this problem, where I had deprecated iris.fileformats.ff.
I think it relates to : http://stackoverflow.com/a/17705456
**Does anyone know how to fix this properly without having to define an independent ```__init__``` for each wrapper class ?**

Comment (rhattersley) [inline lib/iris/_deprecation_helpers.py]: > how to fix this properly

Use Python 3.4 :wink: It's [more complicated](http://stackoverflow.com/a/147878) for Python 2.

File: lib/iris/_deprecation_helpers.py (hunk 2 above)
Comment (pp-mo) [inline lib/iris/_deprecation_helpers.py]: No, but after a bad time with Sphinx I think I'm reworking all of this now ...

**Comment by pp-mo** [comment]
Thanks @rhattersley 
I hope this addresses the outstanding.
I have given up on automatically generating a wrapped ```__init__``` function for wrapper classes, as Sphinx seems to take its constructor signature from any overriding constructor function -- either ```__init__``` or ```__new__```, so I couldn't find any way of making it automatically document the signature of the wrapped ```__init__``` method.
The result is ugly, but it's only "temporary" code.

**Comment by pp-mo** [comment]
P.S.
> documentation is not getting the constructor signatures right

Just to be clear, the point at issue here is not any oddness of Python, or metaclasses, but an oddness of Sphinx. I can fix "help(x)", but not the Sphinx output, apparently because it defines an overall "class constructor signature" in a non-obvious way.
I think you can probably only really work out what it does by trying it.
Sadly, ***that*** isn't odd at all but almost standard practice in that area ...

**Comment by rhattersley** [comment]
> Just to be clear, the point at issue here is not any oddness of Python, or metaclasses, but an oddness of Sphinx.

Thanks for clarifying. :+1: Given the deprecated classes are transient code and you've already boiled it down to a simple, repeatable pattern I'm OK with leaving your solution as it stands. There's no need to burn extra time on a "fancy" solution that works with every permutation of Python 2/3 and Sphinx.

**Comment by rhattersley** [comment]
@pp-mo - I'm :+1: for the content. With hindsight, can we easily squash this down to a small number (but more than one) of coherent commits? If not, I think I'd rather merge it as it stands than squash down to a single commit.

**Comment by rhattersley** [comment]
> can we easily squash this down to a small number (but more than one) of coherent commits?

Nudge...?

**Comment by pp-mo** [comment]
> Nudge...?

Thanks @rhattersley. Coming...

**Comment by pp-mo** [comment]
> can we easily squash this down to a small number (but more than one) of coherent commits?

Now done that, or something like it.
But shortly, no, it was a total beast to do !!

**Comment by rhattersley** [comment]
Thanks @pp-mo :+1: :smile:

**Comment by pelson** [comment]
Just following this up now (I know, how many months ago), but I'm not sure it is clear what functionality has replaced `nearest_neighbour_indices`. I guess the problem with this "scheme" is that it isn't really an interpolation at all - it is a "find the 'nearest' point (Euclidean distance) and tell me what index I should use to get it". It is actually an extremely useful function that I'd be sad to see gone. @pp-mo - as I say, sorry to drag this back up, but could you please advise?

Thanks!

**Comment by pp-mo** [comment]
> what functionality has replaced nearest_neighbour_indices

There's a redirect in the docstring for the deprecated routine.  It says ...
> Please replace usage of iris.analysis.interpolate.nearest_neighbour_indices() with iris.coords.Coord.nearest_neighbour_index().

***But*** I just realised I messed up the docs build for this in the way I wrapped these routines, so we are no longer building docs for them (!oops)
#2270 addresses

**Comment by pp-mo** [comment]
P.S. 
> Please replace usage of iris.analysis.interpolate.nearest_neighbour_indices() with iris.coords.Coord.nearest_neighbour_index().

Does this actually address the need you mention, or am I maybe missing something about the usage with multiple sample-point requirements ?

E.G. Here's my equivalent code for the docstring example....
```
>>> cube = iris.load_cube(iris.sample_data_path('ostia_monthly.nc'))
>>> sample_points = [('latitude', 0), ('longitude', 10)]
>>> indices = nearest_neighbour_indices(cube, sample_points)
>>> indices
(slice(None, None, None), 9, 12)
>>> 
>>> sample_points_dict = dict(sample_points)
>>> indices_2 = tuple([dimcoord.nearest_neighbour_index(sample_points_dict[dimcoord.name()])
...     if dimcoord.name() in sample_points_dict else slice(None)
...     for dimcoord in cube.coords(dim_coords=True)])
>>> 
>>> indices_2 == indices
True
>>> 
```
If that's not too horrible ?

**Comment by pelson** [comment]
Thanks for your example @pp-mo. Definitely less clear than it was, but I can live with it :smile:

**Pull request merged**